domain/src/main/java/org/oppia/android/domain/spotlight/SpotlightStateController.kt

```diff
@@ -0,0 +1,182 @@
+import org.oppia.android.app.model.SpotlightStateDatabase
+import org.oppia.android.app.model.SpotlightViewState
+import org.oppia.android.data.persistence.PersistentCacheStore
+import org.oppia.android.domain.oppialogger.OppiaLogger
+import org.oppia.android.util.data.AsyncResult
+import org.oppia.android.util.data.DataProvider
+import org.oppia.android.util.data.DataProviders
+import org.oppia.android.util.data.DataProviders.Companion.transformAsync
+import javax.inject.Inject
+import javax.inject.Singleton
+
+private const val CACHE_NAME = "spotlight_checkpoint_database"
+private const val RECORD_SPOTLIGHT_CHECKPOINT_DATA_PROVIDER_ID =
+  "record_spotlight_checkpoint_provider_id"
+private const val RETRIEVE_SPOTLIGHT_CHECKPOINT_DATA_PROVIDER_ID =
+  "retrieve_spotlight_checkpoint_provider_id"
+
+/** Handles saving and retrieving feature spotlight states. */
+@Singleton
+class SpotlightStateController @Inject constructor(
+  private val cacheStoreFactory: PersistentCacheStore.Factory,
+  private val oppiaLogger: OppiaLogger,
+  private val dataProviders: DataProviders,
+) {
+  // thrown when spotlight feature is not set while retrieving or marking spotlight view states
+  class SpotlightFeatureNotFoundException(message: String) : Exception(message)
+
+  private val cacheStoreMap =
+    mutableMapOf<ProfileId, PersistentCacheStore<SpotlightStateDatabase>>()
+
+  /**
+   * Marks the [SpotlightViewState] of a spotlit feature for a given profile as seen
+   *
+   * @param profileId the ID of the profile viewing the spotlight
+   * @param feature the spotlight feature who's view state is to be recorded
+   * @throws SpotlightFeatureNotFoundException when feature is not set correctly
+   */
+  fun markSpotlightViewed(
+    profileId: ProfileId,
+    feature: Spotlight.FeatureCase,
+  ): DataProvider<Any?> {
+    val deferred = recordSpotlightStateAsync(
+      profileId,
+      feature,
+      SpotlightViewState.SPOTLIGHT_SEEN
+    )
+    return dataProviders.createInMemoryDataProviderAsync(
+      RECORD_SPOTLIGHT_CHECKPOINT_DATA_PROVIDER_ID
+    ) {
+      return@createInMemoryDataProviderAsync AsyncResult.Success(deferred.await())
+    }
+  }
+
+  /**
+   * Retrieves the current [SpotlightViewState] of a spotlit feature for a given profile.
+   *
+   * @param profileId the ID of the profile that will be viewing the spotlight
+   * @param feature the spotlight feature to be spotlit
+   *
+   * @return DataProvider containing the current [SpotlightViewState] corresponding to the specified [feature]
+   */
+  fun retrieveSpotlightViewState(
+    profileId: ProfileId,
+    feature: Spotlight.FeatureCase,
+  ): DataProvider<SpotlightViewState> {
+    return retrieveCacheStore(profileId)
+      .transformAsync(
+        RETRIEVE_SPOTLIGHT_CHECKPOINT_DATA_PROVIDER_ID
+      ) {
+
+        val viewState = when (feature) {
+          Spotlight.FeatureCase.ONBOARDING_NEXT_BUTTON -> {
+            it.onboardingNextButton
+          }
```

Comment:
I suggest directly importing the feature case so that this can neatly fit on one line. E.g.:

```suggestion
          ONBOARDING_NEXT_BUTTON -> it.onboardingNextButton
```

Ditto for the other cases, and for the setter cases below.

Reply:
done